build_tools/cmake/build_android.sh

```diff
@@ -29,22 +29,146 @@ else
+
+TARGET="iree-minimal-invoke-module"
+
+echo "Building target ${TARGET} for device"
+echo "------------"
+"${CMAKE_BIN}" --build "${BUILD_ANDROID_DIR}" --target "${TARGET}" -- -k 0
+
+TARGET_BINARY="${BUILD_ANDROID_DIR}/tools/${TARGET}"
+
+TOOLCHAIN_TOOLS_DIR="${ANDROID_NDK}/toolchains/llvm/prebuilt/linux-x86_64/bin"
+NM_TOOL="${TOOLCHAIN_TOOLS_DIR}/llvm-nm"
+STRIP_TOOL="${TOOLCHAIN_TOOLS_DIR}/llvm-strip"
+NM_ARGS="-S --radix=d"
+NM_OUTPUT="$(${NM_TOOL} ${NM_ARGS} ${TARGET_BINARY})"
+
+# If we ever start stripping binaries in the Release build type, that will cause
+# nm to report "no symbols" as a 1-line output but nm doesn't return an error
+# code in that case.
+NM_OUTPUT_LINES="$(wc -l <<< "${NM_OUTPUT}")"
+if (( NM_OUTPUT_LINES < 100 ))
+then
+  echo "FATAL error: unexpectedly short nm output:"
+  echo "${NM_OUTPUT}"
+  exit 1
+fi
+
+NM_OUTPUT_ALL_CXX_SYMBOLS="$(grep '\b_Z' <<< "${NM_OUTPUT}" || true)"
+# The standard C library we're linking against has a snprintf implementation
+# that internally calls a C++ snprintf. As a result we get those C++ snprintf
+# symbols from plain C functions in .c files doing plain snprintf calls.
+# Example: iree_status_format (status.c) -> snprintf -> C++ snprintf.
+# We have no choice but to tolerate that and it's not a problem anyway as this
+# C++ code isn't dragging in any other symbol.
+NM_OUTPUT_CXX_SYMBOLS="$(grep -v '\b_ZL8snprintf' <<< "${NM_OUTPUT_ALL_CXX_SYMBOLS}"|| true)"
+
+if [[ ! -z "${NM_OUTPUT_CXX_SYMBOLS}" ]]
+then
+  CXX_SYMBOLS_SIZE="$(echo "${NM_OUTPUT_CXX_SYMBOLS}" | awk '{sum += $2} END {print sum}')"
+  echo "*********************************************************************"
+  echo "FATAL error: target ${TARGET} contains C++ symbols."
+  echo "*********************************************************************"
+  echo
+  echo "Total size of C++ symbols: ${CXX_SYMBOLS_SIZE} bytes"
+  echo
+  echo "Output of nm ${NM_ARGS} for these C++ symbols:"
+  # We could have `nm` demangle right away with -C, but that would make it
+  # harder to tell which symbols are C++. We would have to run `nm` twice.
+  echo "${NM_OUTPUT_CXX_SYMBOLS}" | c++filt
+  echo
+  exit 1
+fi
 
+${STRIP_TOOL} ${TARGET_BINARY}
+
+TARGET_BINARY_SIZE="$(size -A ${TARGET_BINARY} | grep Total | grep -o '[0-9]\+')"
+echo "Sections size of ${TARGET_BINARY}: ${TARGET_BINARY_SIZE} bytes"
+
+# As of October 2022, the current size is 397827, so this gives 20% growth room.
+TARGET_BINARY_SIZE_LIMIT=480000
+
+if ((TARGET_BINARY_SIZE > TARGET_BINARY_SIZE_LIMIT ))
+then
+  echo "*********************************************************************"
+  echo "FATAL error: ${TARGET_BINARY} exceeds the current arbitrary limit."
+  echo "*********************************************************************"
+  echo
+  echo "Actual size:     ${TARGET_BINARY_SIZE}"
+  echo "Arbitrary limit: ${TARGET_BINARY_SIZE_LIMIT}"
+  echo
+  echo "If this is normal, feel free to bump the arbitrary limit."
```

Comment:
Are there open source size monitoring tools / CI integrations we could use? If it's useful, I'd rather we find a way to more generally support it. We care about code size on many platforms (notably riscv / bare-metal, web, Android, iOS), so it would be nice to reuse the scripting and monitoring/logging/tracking.